web/server/codechecker_server/task_executors/task_manager.py

```diff
@@ -0,0 +1,229 @@
+
+MAX_TOKEN_RANDOM_RETRIES = 10
+
+LOG = get_logger("server")
+
+
+class ExecutorInProgressShutdownError(Exception):
+    """
+    Exception raised to indicate that the background executors are under
+    shutdown.
+    """
+    def __init__(self):
+        super().__init__("Task executor is shutting down!")
+
+
+class TaskManager:
+    """
+    Handles the creation of "Task" status objects in the database and pushing
+    in-memory `AbstractTask` subclass instances to a `Queue`.
+
+    This class is instantiatied for EVERY WORKER separately, and is not a
+    shared resource!
+    """
+
+    def __init__(self, q: Queue, config_db_session_factory,
+                 executor_kill_flag: Value, machine_id: str):
+        self._queue = q
+        self._database_factory = config_db_session_factory
+        self._is_shutting_down = executor_kill_flag
+        self._machine_id = machine_id
+
+    @property
+    def machine_id(self) -> str:
+        """Returns the ``machine_id`` the instance was constructed with."""
+        return self._machine_id
+
+    def allocate_task_record(self, kind: str, summary: str,
+                             user_name: Optional[str],
+                             product: Optional[Product] = None) -> str:
+        """
+        Creates the token and the status record for a new task with the given
+        initial metadata.
+
+        Returns the token of the task, which is a unique identifier of the
+        allocated record.
+        """
+        try_count: int = 0
+        while True:
+            with DBSession(self._database_factory) as session:
+                try:
+                    token = generate_random_token(DBTask._token_length)
+
+                    task = DBTask(token, kind, summary, self.machine_id,
+                                  user_name, product)
+                    session.add(task)
+                    session.commit()
+
+                    return token
+                except sqlalchemy.exc.IntegrityError as ie:
+                    # The only failure that can happen is the PRIMARY KEY's
+                    # UNIQUE violation, which means we hit jackpot by
+                    # generating an already used token!
+                    try_count += 1
+
+                    if try_count >= MAX_TOKEN_RANDOM_RETRIES:
+                        raise KeyError(
+                            "Failed to generate a unique ID for task "
+                            f"{kind} ({summary}) after "
+                            f"{MAX_TOKEN_RANDOM_RETRIES} retries!") from ie
+
+    def create_task_data(self, token: str) -> Path:
+        """
+        Creates a temporary directory which is **NOT** cleaned up
+        automatically, and suitable for putting arbitrary files underneath
+        to communicate large inputs (that should not be put in the `Queue`)
+        to the `execute` method of an `AbstractTask`.
+        """
+        task_tmp_root = Path(tempfile.gettempdir()) / "codechecker_tasks" \
```

Comment:
It's not a real question, I'm just thinking loudly: we put things under `/tmp`, because it's server side, right? Only the client is using the workspace directory instead of `/tmp` for temporary files, right?

Reply:
Yes. This is all server-side. Just a temporary directory that's more identifiable than a truly "random" temporary directory for the task's tmp data (like the unzip of a mass-store-run). Because servers can fail in a way that the cleaning of temp directories are not executed, as we are no longer using the tempfile... context managers here.